crates/ol/stf/src/verification.rs

```diff
@@ -0,0 +1,202 @@
+    for (msg, proof) in msg_proofs {
+        let msg_bytes: Vec<u8> = msg.as_ssz_bytes();
+        let hash = StrataHasher::hash_leaf(&msg_bytes);
+        if !state.inbox_mmr().verify(proof, &hash) {
+            return Err(StfError::InvalidMessageProof {
+                account_id,
+                message: msg.clone(),
+            });
+        }
+    }
+    Ok(())
+}
+
+pub(crate) fn verify_update_witness<S: StateAccessor>(
+    sender: AccountId,
+    acct_state: &<S as StateAccessor>::AccountState,
+    update: &SnarkAccountUpdate,
+    _state_accessor: &S,
+) -> StfResult<()> {
+    let snark_state = match acct_state.get_type_state()? {
+        AccountTypeState::Empty => return Ok(()),
+        AccountTypeState::Snark(state) => state,
+    };
+    let vk = snark_state.verifier_key();
+    let claim: Vec<u8> = compute_update_claim::<S>(acct_state, update.operation());
+    let is_valid = vk
+        .verify_claim_witness(&claim, update.update_proof())
+        .is_ok();
+
+    if !is_valid {
+        return Err(StfError::InvalidUpdateProof { account_id: sender });
+    }
+
+    Ok(())
+}
+
+fn compute_update_claim<S: StateAccessor>(
+    _acct_state: &<S as StateAccessor>::AccountState,
+    _operation: &UpdateOperationData,
+) -> Vec<u8> {
+    // Use new state, processed messages, old state, refs and outputs to compute claim
+    todo!()
+}
+
+pub(crate) fn verify_update_outputs_safe<S: StateAccessor>(
+    outputs: &UpdateOutputs,
+    sender: AccountId,
+    acct_state: &S::AccountState,
+    state_accessor: &S,
+) -> StfResult<()> {
+    let original_balance = acct_state.balance();
+    let transfers = outputs.transfers();
+    let messages = outputs.messages();
+
+    // Check if sender exists
+    if !state_accessor.check_account_exists(sender)? {
+        return Err(StfError::NonExistentAccount(sender));
+    }
+
+    // Check if receivers exist
+    for t in transfers {
+        if !state_accessor.check_account_exists(t.dest())? {
+            return Err(StfError::NonExistentAccount(t.dest()));
+        }
+    }
+
+    for m in messages {
+        if !state_accessor.check_account_exists(m.dest())? {
+            return Err(StfError::NonExistentAccount(m.dest()));
+        }
+    }
+
+    let total_sent = outputs
+        .total_output_value()
+        .ok_or(StfError::BitcoinAmountOverflow)?;
+
+    // Check if there is sufficient balance.
+    if total_sent > original_balance {
+        return Err(StfError::InsufficientBalance);
+    }
+    Ok(())
+}
+
+pub(crate) fn verify_ledger_refs<S: StateAccessor>(
+    _ledger_refs: &LedgerRefs,
+    _state_accessor: &S,
+) -> StfResult<()> {
+    // TODO: implement this
+    Ok(())
+}
+
+/// Type safe update that indicates it has been verified.
+#[derive(Clone, Debug)]
+pub(crate) struct VerifiedUpdate<'a> {
+    operation: &'a UpdateOperationData,
+}
+
+impl<'a> VerifiedUpdate<'a> {
+    pub(crate) fn operation(&self) -> &'a UpdateOperationData {
+        self.operation
+    }
+}
```

Comment:
This isn't quite as useful as it seems when we have to call `.operation()` on it and get the "maybe unverified update" everywhere we use it.